Src/Support/Google.Apis.Auth.Tests/OAuth2/ImpersonatedCredentialTests.cs

```diff
@@ -0,0 +1,230 @@
+/*
+Copyright 2021 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Auth.OAuth2.Flows;
+using Google.Apis.Auth.OAuth2.Responses;
+using Google.Apis.Json;
+using Google.Apis.Tests.Mocks;
+using Google.Apis.Util;
+using Moq;
+using System;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Apis.Auth.Tests.OAuth2
+{
+    public class ImpersonatedCredentialTests
+    {
+        private IClock clock = new MockClock(new DateTime(2020, 5, 13, 15, 0, 0, 0, DateTimeKind.Utc));
+
+        private string errorResponseContent = "{\"error\": {\"code\": 404, \"message\": \"...\", \"status\": \"NOT_FOUND\"}}";
+
+        internal class MockHttpMessageHandler : HttpMessageHandler
+        {
+            private string content;
+            private HttpStatusCode statusCode;
+
+            public MockHttpMessageHandler(HttpStatusCode statusCode, string content) {
+                this.content = content;
+                this.statusCode = statusCode;
+            }
+
+            protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+            {
+                HttpResponseMessage response = new HttpResponseMessage(this.statusCode)
+                {
+                    Content = new StringContent(this.content),
+                    RequestMessage = request
+                };
+                return Task.FromResult(response);
+            }
+        }
+
+        private GoogleCredential CreateSourceCredential()
+        {
+            var tokenResponse = new TokenResponse
+                {
+                    AccessToken = "ACCESS_TOKEN",
+                    ExpiresInSeconds = 60 * 10,
+                    IssuedUtc = clock.UtcNow,
+                    RefreshToken = "REFRESH_TOKEN",
+                };
+            var flowMock = new Mock<IAuthorizationCodeFlow>(MockBehavior.Strict);
```

Comment:
It feels odd to have a strict mock for this. Can't we create a simple stub? (Do we really care if the implementation fetches the `Clock` property multiple times, for example?)

Reply:
Here we want to create a mock user credential and set the clock properly so the access token from its token response will be used (i.e. no refresh).

Reply:
I see. Now it uses a class instead of a mock.